src/docfx/build/toc/parser/TableOfContentsParser.cs

```diff
@@ -116,31 +121,58 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
             Document rootPath,
             ResolveContent resolveContent,
             ResolveHref resolveHref,
-            ResolveXref resolveXref)
+            ResolveXref resolveXref,
+            ResolveMoniker resolveMoniker)
         {
             var errors = new List<Error>();
             foreach (var tocModelItem in tocModelItems)
             {
                 if (tocModelItem.Items != null && tocModelItem.Items.Any())
                 {
-                    errors.AddRange(ResolveTocModelItems(context, tocModelItem.Items, parents, filePath, rootPath, resolveContent, resolveHref, resolveXref));
+                    errors.AddRange(ResolveTocModelItems(context, tocModelItem.Items, parents, filePath, rootPath, resolveContent, resolveHref, resolveXref, resolveMoniker));
                 }
 
                 // process
                 var tocHref = GetTocHref(tocModelItem);
                 var topicHref = GetTopicHref(tocModelItem);
                 var topicUid = tocModelItem.Uid;
 
+                var monikers = new List<string>();
+
                 var (resolvedTocHref, resolvedTopicItemFromTocHref, subChildren) = ProcessTocHref(tocHref);
                 var (resolvedTopicHref, resolvedTopicName, document) = ProcessTopicItem(topicUid, topicHref);
 
                 // set resolved href back
                 tocModelItem.Href = resolvedTocHref ?? resolvedTopicHref ?? resolvedTopicItemFromTocHref?.Href;
+                if (resolvedTocHref != null || resolvedTopicHref != null)
+                {
+                    var linkType = UrlUtility.GetLinkType(resolvedTopicHref);
+                    if (linkType == LinkType.External || linkType == LinkType.AbsolutePath)
+                    {
+                        monikers = resolveMoniker(rootPath);
+                    }
+                    else
+                    {
+                        monikers = resolveMoniker(document);
+                    }
+                }
+                else
+                {
+                    tocModelItem.Href = resolvedTopicItemFromTocHref?.Href;
+                    monikers = resolvedTopicItemFromTocHref?.Monikers ?? new List<string>();
+                }
+
                 tocModelItem.TocHref = resolvedTocHref;
                 tocModelItem.Homepage = resolvedTopicHref;
                 tocModelItem.Name = tocModelItem.Name ?? resolvedTopicName;
                 tocModelItem.Items = subChildren?.Items ?? tocModelItem.Items;
 
+                // Union with children's monikers
+                var childrenMonikers = tocModelItem.Items?.SelectMany(child => child?.Monikers ?? new List<string>()) ?? new List<string>();
+                monikers = childrenMonikers.Union(monikers).Distinct().ToList();
+                monikers.Sort(context.MonikerProvider.Comparer);
+                tocModelItem.Monikers = monikers;
```

Comment:
consider extract a method to contain logics in line 170-174 and line 147-165
```suggestion
                tocModelItem.Monikers = GetMonikers(xxxxx);
```